File: app/src/main/java/com/pomo/sync/transport/ProviderWrap.kt
package com.pomo.sync.transport

import com.pomo.sync.crypto.AesGcmCiphertext
import com.pomo.sync.crypto.PomoCrypto
import com.pomo.sync.protocol.CborValue
import com.pomo.sync.protocol.DeterministicCbor

/**
 * Optional content-epoch wrapping for provider object bytes (WebDAV / Nostr).
 * When no key is installed, bytes pass through unchanged.
 */
internal object ProviderWrap {
    const val LABEL: String = "pomo-provider-wrapped"
    const val SCHEMA: Long = 1

    @Volatile
    private var contentKey: ByteArray? = null

    fun installContentKey(key: ByteArray?) {
        require(key == null || key.size == 32)
        contentKey = key?.copyOf()
    }

    fun installed(): Boolean = contentKey != null

    fun wrap(plaintext: ByteArray): ByteArray {
        val key = contentKey ?: return plaintext.copyOf()
        val nonce = ByteArray(12).also { java.security.SecureRandom().nextBytes(it) }
        val sealed = PomoCrypto.encryptAesGcm(key, nonce, ByteArray(0), plaintext)
        return DeterministicCbor.encode(
            CborValue.Array(
                listOf(
                    CborValue.Text(LABEL),
                    CborValue.Integer(SCHEMA),
                    CborValue.Bytes(sealed.nonce.copyOf()),
                    CborValue.Bytes(sealed.ciphertextAndTag.copyOf()),
                ),
            ),
        )
    }

    fun open(bytes: ByteArray): ByteArray {
        val key = contentKey ?: return bytes.copyOf()
        val fields =
            runCatching {
                (DeterministicCbor.decodeCanonical(bytes) as? CborValue.Array)?.values
            }.getOrNull() ?: return bytes.copyOf()
        if (fields.size != 4 || fields[0] != CborValue.Text(LABEL) || fields[1] != CborValue.Integer(SCHEMA)) {
            return bytes.copyOf()
        }
        val nonce = (fields[2] as? CborValue.Bytes)?.value ?: error("invalid provider wrap nonce")
        val ciphertext = (fields[3] as? CborValue.Bytes)?.value ?: error("invalid provider wrap ciphertext")
        return PomoCrypto.decryptAesGcm(key, AesGcmCiphertext(nonce, ciphertext), ByteArray(0))
    }
}

File: extension/src/sync/transport/providerWrap.ts
import { decryptAes256Gcm, encryptAes256Gcm } from "../crypto/PomoCrypto";
import { encodeCanonicalCbor, decodeCanonicalCbor, type CborValue } from "../protocol/cbor";

const LABEL = "pomo-provider-wrapped";
const SCHEMA = 1;

let activeContentKey: Uint8Array | null = null;

/** Install the current content-epoch key used to wrap provider object bytes. */
export function installProviderContentKey(key: Uint8Array | null): void {
  if (key !== null && key.length !== 32) throw new Error("provider content key must be 32 bytes");
  activeContentKey = key === null ? null : key.slice();
}

export function providerContentKeyInstalled(): boolean {
  return activeContentKey !== null;
}

export async function wrapProviderBytes(plaintext: Uint8Array): Promise<Uint8Array> {
  const key = activeContentKey;
  if (key === null) return plaintext.slice();
  const sealed = await encryptAes256Gcm(key, crypto.getRandomValues(new Uint8Array(12)), new Uint8Array(), plaintext);
  return encodeCanonicalCbor([LABEL, SCHEMA, sealed.nonce, sealed.ciphertextAndTag]);
}

export async function openProviderBytes(bytes: Uint8Array): Promise<Uint8Array> {
  const key = activeContentKey;
  if (key === null) return bytes.slice();
  let fields: readonly CborValue[];
  try {
    const decoded = decodeCanonicalCbor(bytes);
    if (!Array.isArray(decoded) || decoded.length !== 4 || decoded[0] !== LABEL || decoded[1] !== SCHEMA) {
      return bytes.slice();
    }
    fields = decoded;
  } catch {
    return bytes.slice();
  }
  const nonce = fields[2];
  const ciphertext = fields[3];
  if (!(nonce instanceof Uint8Array) || !(ciphertext instanceof Uint8Array)) throw new Error("invalid provider wrap");
  return decryptAes256Gcm(key, { nonce, ciphertextAndTag: ciphertext }, new Uint8Array());
}
